### PR TITLE
[GOBBLIN-391] Use the DataPublisherFactory to allow sharing publisher…

### PR DESCRIPTION
**Pull request by htran1**
…s in SafeDatasetCommit

Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [X] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-391

### Description
- [X] Here are some details about my PR, including screenshots (if applicable):
Some publishers are threadsafe and are shareable. Share these when possible to reduce instantiation cost.

### Tests
- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [X] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by htran1** [comment]
@ibuenros, please review.

**Comment by htran1** [comment]
@ibuenros, please review.

**Comment by htran1** [comment]
@ibuenros, please review

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/SafeDatasetCommit.java [hunk 1]
```diff
@@ -133,9 +136,21 @@ public Void call()
...
-                : taskFactory.createDataPublisher(this.datasetState);
+            DataPublisher publisher;
+
+            if (taskFactory == null) {
+              publisher = DataPublisherFactory.get(dataPublisherClass.getName(), this.jobContext.getJobState(),
+                  this.jobContext.getJobBroker());
+
+              // non-threadsafe publishers are not shareable and are not retained in the broker, so register them with
+              // the closer
+              if (!publisher.supportsCapability(Capability.THREADSAFE, Collections.EMPTY_MAP)) {
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/SafeDatasetCommit.java]: Can you move this logic to a method in `DataPublisherFactory`?

File: gobblin-runtime/src/main/java/org/apache/gobblin/runtime/SafeDatasetCommit.java [hunk 2]
```diff
@@ -133,9 +136,21 @@ public Void call()
...
+                closer.register(publisher);
+              }
+            } else {
+              publisher = taskFactory.createDataPublisher(this.datasetState);
```
Comment (ibuenros) [inline gobblin-runtime/src/main/java/org/apache/gobblin/runtime/SafeDatasetCommit.java]: Please add a TODO to remember we may have to solve the shared publisher issue for task factories.

**Pull request closed without merging**